File: sdk/js/node/src/credentials.ts
import type { HttpClient } from "./client.js";

export interface Credential {
  id: string;
  subject: string;
  type: string;
  issued_at: string;
  expires_at?: string;
  revoked: boolean;
}

export interface IssueCredentialInput {
  subject: string;
  type: string;
  claims?: Record<string, unknown>;
  ttl_seconds?: number;
}

export interface IssueCredentialResult {
  credential_id: string;
  jwt: string;
  expires_at?: string;
}

export interface VerifyCredentialResult {
  valid: boolean;
  reason?: string;
  subject?: string;
  issuer?: string;
  vc?: Record<string, unknown>;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
}

export class Credentials {
  constructor(private readonly http: HttpClient) {}

  /** Issue a new W3C JWT-VC for a subject. */
  issue(tenantId: string, input: IssueCredentialInput): Promise<IssueCredentialResult> {
    return this.http.post<IssueCredentialResult>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/credentials`,
      input,
    );
  }

  /** List issued credentials for a tenant. */
  async list(tenantId: string): Promise<Credential[]> {
    const res = await this.http.get<ListEnvelope<Credential>>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/credentials`,
    );
    return res.items ?? res.data ?? [];
  }

  /** Revoke a credential. */
  revoke(tenantId: string, id: string): Promise<void> {
    return this.http.post<void>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/credentials/${encodeURIComponent(id)}/revoke`,
      {},
    );
  }

  /**
   * Verify a presented JWT-VC. Public endpoint — no API key required.
   * Relying parties call this to confirm authenticity + revocation status.
   */
  verify(jwt: string): Promise<VerifyCredentialResult> {
    return this.http.post<VerifyCredentialResult>("/v1/credentials/verify", {
      credential: jwt,
    });
  }
}

File: sdk/go/credentials.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
)

type Credential struct {
	ID        string `json:"id"`
	Subject   string `json:"subject"`
	Type      string `json:"type"`
	IssuedAt  string `json:"issued_at"`
	ExpiresAt string `json:"expires_at,omitempty"`
	Revoked   bool   `json:"revoked"`
}

type IssueCredentialInput struct {
	Subject    string         `json:"subject"`
	Type       string         `json:"type"`
	Claims     map[string]any `json:"claims,omitempty"`
	TTLSeconds int            `json:"ttl_seconds,omitempty"`
}

type IssueCredentialResult struct {
	CredentialID string `json:"credential_id"`
	JWT          string `json:"jwt"`
	ExpiresAt    string `json:"expires_at,omitempty"`
}

type VerifyCredentialResult struct {
	Valid   bool           `json:"valid"`
	Reason  string         `json:"reason,omitempty"`
	Subject string         `json:"subject,omitempty"`
	Issuer  string         `json:"issuer,omitempty"`
	VC      map[string]any `json:"vc,omitempty"`
}

type Credentials struct{ http *httpClient }

func (r *Credentials) Issue(ctx context.Context, tenantID string, in IssueCredentialInput) (*IssueCredentialResult, error) {
	var out IssueCredentialResult
	err := r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/credentials", nil, in, &out, false)
	return &out, err
}

func (r *Credentials) List(ctx context.Context, tenantID string) ([]Credential, error) {
	var env struct {
		Items []Credential `json:"items"`
		Data  []Credential `json:"data"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/tenants/"+url.PathEscape(tenantID)+"/credentials", nil, nil, &env, true); err != nil {
		return nil, err
	}
	if env.Items != nil {
		return env.Items, nil
	}
	return env.Data, nil
}

func (r *Credentials) Revoke(ctx context.Context, tenantID, id string) error {
	return r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/credentials/"+url.PathEscape(id)+"/revoke", nil, struct{}{}, nil, false)
}

// Verify is the public endpoint — no API key required. Relying parties call
// this to confirm a presented JWT-VC is authentic and not revoked.
func (r *Credentials) Verify(ctx context.Context, jwt string) (*VerifyCredentialResult, error) {
	var out VerifyCredentialResult
	body := map[string]string{"credential": jwt}
	err := r.http.do(ctx, http.MethodPost, "/v1/credentials/verify", nil, body, &out, false)
	return &out, err
}
